go/libraries/doltcore/sqle/dolt_index.go

```diff
@@ -151,31 +166,33 @@ func (di *doltIndex) Table() string {
 	return di.tableName
 }
 
-func (di *doltIndex) keysToExpr(keys []interface{}, compFunc compExpr) sql.Expression {
-	exprs := make([]sql.Expression, len(keys))
-	for i := 0; i < len(keys); i++ {
-		exprs[i] = compFunc(
-			expression.NewGetField(i, di.cols[i].TypeInfo.ToSqlType(), di.cols[i].Name, di.cols[i].IsNullable()),
-			expression.NewLiteral(keys[i], di.cols[i].TypeInfo.ToSqlType()),
-		)
-	}
-	lastExpr := exprs[len(exprs)-1]
-	for i := len(exprs) - 2; i >= 0; i-- {
-		lastExpr = expression.NewAnd(exprs[i], lastExpr)
-	}
-	return lastExpr
+func (di *doltIndex) TableData() types.Map {
+	return di.tableData
 }
 
-func (di *doltIndex) keysToIter(expr sql.Expression) (sql.IndexLookup, error) {
-	crf, err := CreateReaderFuncLimitedByExpressions(di.rowData.Format(), di.mapSch, []sql.Expression{expr})
-	if err != nil {
-		return nil, err
+func (di *doltIndex) keysToTuple(keys []interface{}, appendMaxValue bool) (types.Tuple, error) {
+	nbf := di.indexRowData.Format()
+	if len(di.cols) != len(keys) {
+		return types.EmptyTuple(nbf), errors.New("keys must specify all columns for an index")
 	}
-	mapIter, err := crf(di.ctx, di.rowData)
-	if err != nil {
-		return nil, err
+	var vals []types.Value
+	for i, col := range di.cols {
+		val, err := col.TypeInfo.ConvertValueToNomsValue(keys[i])
+		if err != nil {
+			return types.EmptyTuple(nbf), err
+		}
+		vals = append(vals, types.Uint(col.Tag), val)
+	}
+	// In the case of possible partial keys, we may need to match at the beginning or end for matched values, so we
+	// append a tag that is beyond the allowed maximum. This will be ignored if it's a full key and not a partial key.
+	if appendMaxValue {
```

Comment:
I don't think this is right, Brian should take a look.

If you're trying to get a start value for iterating over the map, then you need a start tuple which has the same number of keys as the schema does. You're only including the number of indexed columns, plus maybe one optional value for the first column of the primary key. What you have now only works if the primary keys are all >= the zero value for the underlying noms type. If you had a table with negative primary keys, I believe this code would start iterating in the middle of the map (where keys >= 0). I'm not positive, Brian should say more.

Reply:
So appendMaxValue looks like it's only true for:

DescendGreater - Should be good because it skips the current key and points to position before next possible key.
DescendLessOrEquald - Should be good because it is going past the last key, but is always before the next key.  Will work for partial and full keys.
AscendRange - Used for the end of the range which should be good assuming range is inclusive.  In this case it is the same as above.

Reply:
Double check your test cases to make sure Zach's concerns are covered.